File: crates/shared/src/models/websites.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::{Error, FromRow, Row, postgres::PgRow, types::Json};
use sqlx_pg_ext_uint::c_u16::U16;
use url::Url;

use crate::objectid::ObjectId;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseWebsite {
    pub id: ObjectId,
    pub name: Option<String>,
    pub hosts: Vec<String>,
    pub ports: Vec<u16>,
    pub certificates: Vec<ObjectId>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub backends: Vec<DatabaseWebsiteBackend>,
    pub config: DatabaseWebsiteConfig,
}

impl<'r> FromRow<'r, PgRow> for DatabaseWebsite {
    fn from_row(row: &PgRow) -> Result<Self, Error> {
        Ok(DatabaseWebsite {
            id: row.try_get("id")?,
            name: row.try_get("name")?,
            hosts: row.try_get("hosts")?,
            ports: row
                .try_get::<Vec<U16>, _>("ports")?
                .iter()
                .map(|v| u16::from(*v))
                .collect::<Vec<u16>>(),
            certificates: row.try_get("certificates")?,
            created_at: row.try_get("created_at")?,
            updated_at: row.try_get("updated_at")?,
            backends: row
                .try_get::<Json<Vec<DatabaseWebsiteBackend>>, _>("backends")?
                .0,
            config: row.try_get::<Json<DatabaseWebsiteConfig>, _>("config")?.0,
        })
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DatabaseWebsiteBackend {
    pub url: Url,
    pub balance: usize,
    pub main: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct DatabaseWebsiteConfig {
    pub get_request_ip: DatabaseWebsiteRequestIp,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase", tag = "type", content = "data")]
pub enum DatabaseWebsiteRequestIp {
    #[default]
    Raw,
    ProxyProtocol,
    XForwardedFor(Option<usize>),
    XRealIP,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateDatabaseWebsite {
    pub name: Option<String>,
    pub hosts: Vec<String>,
    pub ports: Vec<u16>,
    pub certificates: Vec<ObjectId>,
    pub backends: Vec<DatabaseWebsiteBackend>,
    pub config: Option<DatabaseWebsiteConfig>,
}

File: dashboard/src/types/websites.ts
export interface WebsiteBackendInput {
    url: string;
    balance: number;
}

export interface Website {
    id: string;
    name: string;
    hosts: string[];
    ports: number[];
    certificates: string[];
    created_at: string;
    updated_at: string;
    backends: WebsiteBackend[];
    config: WebsiteConfig;
}

export interface WebsiteBackend extends WebsiteBackendInput {
    main: boolean;
}

export interface WebsiteConfig {
    get_request_ip: WebsiteRequestIp;
}

export type WebsiteRequestIpType = 'raw' | 'proxyProtocol' | 'xForwardedFor';

export interface WebsiteRequestIp {
    type: WebsiteRequestIpType;
    data?: number;
}

export interface WebsiteCreateRequest {
    name?: string;
    hosts: string[];
    ports: number[];
    certificates: string[];
    backends: WebsiteBackend[];
    config?: WebsiteConfig;
}
